src/main/kotlin/org/jitsi/nlj/transform/node/incoming/VideoParser.kt

```diff
@@ -61,51 +67,78 @@ class VideoParser(
                     val vp8Packet = packetInfo.packet.toOtherType(::Vp8Packet)
                     packetInfo.packet = vp8Packet
                     packetInfo.resetPayloadVerification()
+
+                    if (videoCodecParser !is Vp8Parser) {
+                        resetSources()
+                        videoCodecParser = Vp8Parser(sources, logger)
+                    }
+                }
+                is Vp9PayloadType -> {
+                    val vp9Packet = packetInfo.packet.toOtherType(::Vp9Packet)
+                    packetInfo.packet = vp9Packet
+                    packetInfo.resetPayloadVerification()
+
+                    if (videoCodecParser !is Vp9Parser) {
+                        resetSources()
+                        videoCodecParser = Vp9Parser(sources, logger)
+                    }
+                }
+                else -> {
+                    videoCodecParser = null
                 }
             }
+            videoCodecParser?.parse(packetInfo)
+            packet = packetInfo.packetAs()
         } catch (e: Exception) {
             logger.error("Exception parsing video packet.  Packet data is: " +
-                "${packet.buffer.toHex(packet.offset, Math.min(packet.length, 80))}", e)
+                packet.buffer.toHex(packet.offset, Math.min(packet.length, 80)), e)
             return null
         }
 
-        val videoPacket = packetInfo.packetAs<VideoRtpPacket>()
-        val encodingDesc = findRtpLayerDesc(videoPacket) ?: run {
-            logger.warn("Unable to find encoding matching packet! packet=$videoPacket, " +
-                "encodings=${sources.joinToString(separator = "\n", limit = 1, truncated = "[${sources.size - 1} " +
-                    "more source descriptions omitted]")}")
-            numPacketsDroppedNoEncoding.incrementAndGet()
-            return null
+        if (packet is ParsedVideoPacket) {
+            /* Some codecs mark keyframes in every packet of the keyframe - only count the start of the frame,
+             * so the count is correct. */
+            /* Alternately we could keep track of keyframes we've already seen, by timestamp, but that seems unnecessary. */
+            if (packet.isKeyframe && packet.isStartOfFrame) {
+                logger.cdebug { "Received a keyframe for ssrc ${packet.ssrc} ${packet.sequenceNumber}" }
+                numKeyframes++
+            }
         }
-        videoPacket.qualityIndex = encodingDesc.index
 
         return packetInfo
     }
 
-    private fun findRtpLayerDesc(packet: VideoRtpPacket): RtpLayerDesc? {
-        for (source in sources) {
-            source.findRtpLayerDesc(packet)?.let {
-                return it
-            }
-        }
-        return null
-    }
-
     override fun handleEvent(event: Event) {
         when (event) {
             is SetMediaSourcesEvent -> {
                 sources = event.mediaSourceDescs
+                origSources = sources.copy()
```

Comment:
so `origSources` is really more the "previous" sources, yeah?  Should we keep the _original_ sources, or do we only need the previous ones?

Reply:
It's the most-recent signaled sources.  I'll use that name.